rpcs3/Emu/RSX/Common/BufferUtils.cpp

```diff
@@ -755,39 +733,270 @@ namespace
+				max = _mm256_max_epu16(max, value_with_min_restart);
+				min = _mm256_min_epu16(min, value_with_max_restart);
+				_mm256_storeu_si256(dst_stream++, value_with_max_restart);
+			}
+
+			const __m128i mask_step1 = _mm_set_epi8(
+				0, 0, 0, 0, 0, 0, 0, 0,
+				0xF, 0xE, 0xD, 0xC, 0xB, 0xA, 0x9, 0x8);
+
+			const __m128i mask_step2 = _mm_set_epi8(
+				0, 0, 0, 0, 0, 0, 0, 0,
+				0, 0, 0, 0, 0x7, 0x6, 0x5, 0x4);
+
+			const __m128i mask_step3 = _mm_set_epi8(
+				0, 0, 0, 0, 0, 0, 0, 0,
+				0, 0, 0, 0, 0, 0, 0x3, 0x2);
+
+			__m128i tmp = _mm256_extracti128_si256(min, 1);
+			__m128i min2 = _mm256_castsi256_si128(min);
+			min2 = _mm_min_epu16(min2, tmp);
+			tmp = _mm_shuffle_epi8(min2, mask_step1);
+			min2 = _mm_min_epu16(min2, tmp);
+			tmp = _mm_shuffle_epi8(min2, mask_step2);
+			min2 = _mm_min_epu16(min2, tmp);
+			tmp = _mm_shuffle_epi8(min2, mask_step3);
+			min2 = _mm_min_epu16(min2, tmp);
+
+			tmp = _mm256_extracti128_si256(max, 1);
+			__m128i max2 = _mm256_castsi256_si128(max);
+			max2 = _mm_max_epu16(max2, tmp);
+			tmp = _mm_shuffle_epi8(max2, mask_step1);
+			max2 = _mm_max_epu16(max2, tmp);
+			tmp = _mm_shuffle_epi8(max2, mask_step2);
+			max2 = _mm_max_epu16(max2, tmp);
+			tmp = _mm_shuffle_epi8(max2, mask_step3);
+			max2 = _mm_max_epu16(max2, tmp);
+
+			const u16 min_index = u16(_mm_cvtsi128_si32(min2) & 0xFFFF);
+			const u16 max_index = u16(_mm_cvtsi128_si32(max2) & 0xFFFF);
+
+			return std::make_tuple(min_index, max_index);
+		}
+
+		SSE4_1_FUNC
+		static
+		std::tuple<u16, u16> upload_u16_swapped_sse4_1(const void *src, void *dst, u32 iterations, u16 restart_index)
+		{
+			const __m128i shuffle_mask = _mm_set_epi8(
+				0xE, 0xF, 0xC, 0xD,
+				0xA, 0xB, 0x8, 0x9,
+				0x6, 0x7, 0x4, 0x5,
+				0x2, 0x3, 0x0, 0x1);
+
+			auto src_stream = (const __m128i*)src;
+			auto dst_stream = (__m128i*)dst;
+
+			__m128i restart = _mm_set1_epi16(restart_index);
+			__m128i min = _mm_set1_epi16(0xffff);
+			__m128i max = _mm_set1_epi16(0);
+
+			for (unsigned n = 0; n < iterations; ++n)
+			{
+				const __m128i raw = _mm_loadu_si128(src_stream++);
+				const __m128i value = _mm_shuffle_epi8(raw, shuffle_mask);
+				const __m128i mask = _mm_cmpeq_epi16(restart, value);
+				const __m128i value_with_min_restart = _mm_andnot_si128(mask, value);
+				const __m128i value_with_max_restart = _mm_or_si128(mask, value);
+				max = _mm_max_epu16(max, value_with_min_restart);
+				min = _mm_min_epu16(min, value_with_max_restart);
+				_mm_storeu_si128(dst_stream++, value_with_max_restart);
+			}
+
+			const __m128i mask_step1 = _mm_set_epi8(
+				0, 0, 0, 0, 0, 0, 0, 0,
+				0xF, 0xE, 0xD, 0xC, 0xB, 0xA, 0x9, 0x8);
+
+			const __m128i mask_step2 = _mm_set_epi8(
+				0, 0, 0, 0, 0, 0, 0, 0,
+				0, 0, 0, 0, 0x7, 0x6, 0x5, 0x4);
+
+			const __m128i mask_step3 = _mm_set_epi8(
+				0, 0, 0, 0, 0, 0, 0, 0,
+				0, 0, 0, 0, 0, 0, 0x3, 0x2);
+
+			__m128i tmp = _mm_shuffle_epi8(min, mask_step1);
```

Comment:
Instead of shuffles, can use PSRLDQ to shift data to the right (_mm_srli_si128).
16-bit minimum can also be done with PHMINPOSUW (_mm_minpos_epu16).

Reply:
VPHMINPOSUW indeed lets me economise quite a few instructions, but there is no equivalent for max or for 32-bit, which is quite surprising.